src/gui/widgets/VectorGraph.cpp

```diff
@@ -0,0 +1,777 @@
+		update();
+	}
+
+	if (model()->getCurrentDraggedTensionHandle() != -1)
+	{
+		QCursor c = cursor();
+		float delta = c.pos().y() - model()->getStoredCursorPos().y();
+		c.setPos(model()->getStoredCursorPos());
+		setCursor(c);
+
+		int index = model()->getCurrentDraggedTensionHandle();
+		VectorGraphPoint * previousPoint = model()->getPoint(index - 1);
+		VectorGraphPoint * pointToEdit = model()->getPoint(index);
+
+		if (previousPoint->y() > pointToEdit->y())
+			delta *= -1;
+
+		// Subtracting, moving down vertically makes the y value go up
+		float newTension = pointToEdit->tension() - delta / 250; // Make adjustable from somewhere else - this is an important tweak
+		if (newTension > 1)
+			newTension = 1;
+		else if (newTension < -1)
+			newTension = -1;
+		pointToEdit->setTension(newTension);
+		update();
+	}
+}
+
+void VectorGraph::mouseReleaseEvent(QMouseEvent * event)
+{
+	if (model()->getCurrentDraggedPoint() != -1)
+		model()->resetCurrentDraggedPoint();
+
+	if (model()->getCurrentDraggedTensionHandle() > -1)
+	{
+		VectorGraphPoint * point = model()->getPoint(model()->getCurrentDraggedTensionHandle());
+		model()->setLastModifiedTension(point->tension());
+		model()->setLastModifiedTensionType(point->tensionType());
+
+		QCursor c = cursor();
+		QPoint newCursorPoint(
+					rawToCoordX(getTensionHandleXVal(model()->getCurrentDraggedTensionHandle())),
+					rawToCoordY(getTensionHandleYVal(model()->getCurrentDraggedTensionHandle())));
+		c.setPos(mapToGlobal(newCursorPoint));
+		c.setShape(Qt::ArrowCursor);
+		setCursor(c);
+		model()->resetCurrentDraggedTensionHandle();
+	}
+}
+
+bool VectorGraph::eventFilter(QObject *watched, QEvent *event)
+{
+	if (event->type() == QEvent::ContextMenu)
+	{
+		if (model()->getCurrentDraggedPoint() >= 0)
+			return false;
+
+		QContextMenuEvent * menuEvent = static_cast<QContextMenuEvent*>(event);
+
+		m_currentPoint = model()->getPointIndexFromCoords(coordToRawX(menuEvent->x()) * m_width, coordToRawY(menuEvent->y()) * m_height, m_width, m_height);
+
+		if (m_currentPoint < 0)
+			return false;
+
+		CaptionMenu contextMenu(model()->displayName(), this);
+		contextMenu.addAction(QPixmap(), tr("Hold"), this, SLOT(setTensionToHold()));
+		contextMenu.addAction(QPixmap(), tr("Single Curve"), this, SLOT(setTensionToSingle()));
+		contextMenu.addAction(QPixmap(), tr("Double Curve"), this, SLOT(setTensionToDouble()));
+		contextMenu.addAction(QPixmap(), tr("Stairs"), this, SLOT(setTensionToStairs()));
+		contextMenu.addAction(QPixmap(), tr("Pulse"), this, SLOT(setTensionToPulse()));
+		contextMenu.addAction(QPixmap(), tr("Wave"), this, SLOT(setTensionToWave()));
+		if (model()->getPoint(m_currentPoint)->canBeDeleted())
+		{
+			contextMenu.addSeparator();
+			contextMenu.addAction(QPixmap(), tr("&Delete"), this, SLOT(deletePoint()));
+		}
+		contextMenu.exec(QCursor::pos());
+		return true;
+	}
+	return false;
+}
+
+float VectorGraph::calculateSample(float input)
```

Comment:
Do we need to have this as a separate function? I think one can use `model()->calculateSample` directly.